### ISSUE TITLE
Open Attribute Table -- LIMIT by default to X items

### ISSUE THREAD

**Issue by cochetti** [qgis/QGIS#57835]
### Feature description

1. **Request**  

### Limit the number of rows output by open attribute table by default, add a button to override to show all. 

I frequently like to check the attribute table to see whether the data has written correctly, the a by-location query has joined correctly, to check xyz points, etc. I have noticed the significant overhead of this operation and it has crashed the program a number of times. 

I would like to suggest using the LIMIT option on the SQL query behind the "Open Attribute Table" is activated by default. This is customary in many (if not most) SQL varieties and saves network bandwidth, memory access, RAM, and basically can only help users. There are very few users who would knowingly and without hesitation open a 36,000,000+ point table. I suspect that having a liberal enough limit, but large enough to be representative will meet most users' needs. 

2. **Expanded Feature**  
### Randomly select among points when doing so 

To be perfect certain of that fact the addition of the following query would randomly select points to make sure that if the data are specifically ordered the first ones aren't all drawn from a single corner of the map:   

```
SELECT *
FROM layer
ORDER BY rand()
LIMIT 50000 
```

**Comment by NyakudyaA** [comment]
The attribute toolbar already has options to prevent you from opening a large attribute table
<img width="1080" alt="Screenshot 2024-06-23 at 16 37 07" src="https://github.com/qgis/QGIS/assets/2510900/7996a7a2-4632-413f-8055-3f8dce4166a8">

I think a better solution would be to add the option to the context menu with the options you requesting i.e
* Open the first 100 features


What would be the use case for opening the first 100 random features? Is it to inspect the data only? I feel like most of these options are already available i.e

In processing select *Random Selection*
In the attribute toolbar choose *Open Attribute table Selected Features*

**Comment by cochetti** [comment]
Notice this, the feature you are pointing to only becomes available _after_ opening the massively large table. Thus you don't actually provide protection against doing so until one has already wasted their time waiting for it to load the first time. 

Personally I am used to working in a SQL framework where I can just issue any query and then use a LIMIT qualifier to reduce the number of lines returned. I would really appreciate more control in this manner as there are a lot of times I know what I want to do, however datatype issues that are not as clearly discussed or accessible become a problem. 

For example, it is really frustrating having no indication whether Z data is part of the geometry of a layer or if it is just a column. Certain functions error out badly with no clear indication why because it seems that coordinate system features are inconsistently handled whether or not Z and W boxes are checked on layer creation. 

As a SQL/statistical programmers primarily, when I open the attribute panel more often than not I am trying to get insight into data structure and format. The fact that the geometry is often the only data on a layer, and its completely unviewable is really difficult as s new user. Then add in how Delauney and Voronoi meshes are very happy to spend 10-15 minutes processing and don't give a warning that they are constrained to 2D. I have over months learned the finnicky rules, however it has been a slog. For example, even just the default 4326 is a pain as my data are in projected NAD83 XY using survey feet. When the Z coordinate is present such as embedded in LIDAR data with its ellisoidal system, the inverse transform is not computible and it breaks even the most basic units transformations from say feet to meters as these all rely on PROJ4, but end up with a noproj ()--() error on nearly all operations since they so freely rely on bidirectional transformation. 

Being able to know what the constraints are more easily or having more proactive warnings would prevent a lot of wasted time, program crashes and unresponsive windows, and head banging. Even just a simply icon or shading highlight of the layers to show which have consistent versus incompatible or invertable transform matrices would be game changing .

**Comment by NyakudyaA** [comment]
> Notice this, the feature you are pointing to only becomes available _after_ opening the massively large table. Thus you don't actually provide protection against doing so until one has already wasted their time waiting for it to load the first time.

This is not correct, You do not need to open the attribute table to interact with those options. They become available as soon as you add the layer to the TOC.
> 
> Personally I am used to working in a SQL framework where I can just issue any query and then use a LIMIT qualifier to reduce the number of lines returned. I would really appreciate more control in this manner as there are a lot of times I know what I want to do, however datatype issues that are not as clearly discussed or accessible become a problem.

There are a couple of options here already.

1. Load your vector layer and create a new virtual layer. You can use SQL to limit the returns 
2. If your data is in a database, using the DBBrowser, you can run a SQL statement and load the results as a new layer.
3. If you are using the datasource manager, depending on your datatype you can also filter your data based on SQL and load filtered data.
4. If you have native data types that do not support filtering before loading, you can use the filter options available for the layer using QGIS Expressions and then open the attribute with a subset of features.

**Note:** I am not saying the Limit is not a good feature, I am just pointing out other ways to achieve the same.

> 
> For example, it is really frustrating having no indication whether Z data is part of the geometry of a layer or if it is just a column. Certain functions error out badly with no clear indication why because it seems that coordinate system features are inconsistently handled whether or not Z and W boxes are checked on layer creation.

If you hover over the layer, it will give you a description of the layer, this will indicate the geometry type. For detailed information you can go to the properties and click on Information tab.
> 
> As a SQL/statistical programmers primarily, when I open the attribute panel more often than not I am trying to get insight into data structure and format. The fact that the geometry is often the only data on a layer, and its completely unviewable is really difficult as s new user. Then add in how Delauney and Voronoi meshes are very happy to spend 10-15 minutes processing and don't give a warning that they are constrained to 2D. I have over months learned the finnicky rules, however it has been a slog. For example, even just the default 4326 is a pain as my data are in projected NAD83 XY using survey feet. When the Z coordinate is present such as embedded in LIDAR data with its ellisoidal system, the inverse transform is not computible and it breaks even the most basic units transformations from say feet to meters as these all rely on PROJ4, but end up with a noproj ()--() error on nearly all operations since they so freely rely on bidirectional transformation.

I guess you need to open issue reports for those algorithms you are using and maybe they can be fixed.
> 
> Being able to know what the constraints are more easily or having more proactive warnings would prevent a lot of wasted time, program crashes and unresponsive windows, and head banging. Even just a simply icon or shading highlight of the layers to show which have consistent versus incompatible or invertable transform matrices would be game changing .

Same hee, you can open reports to suggest improvements and if a developer catches this , it will be tackled otherwise you might have to go through 2 approaches

1. Find a developer to implement the suggested changes
2. Develop a plugin that enhances these features or look for existing features

**Comment by antoniolocandro** [comment]
Recently started working with some huget tables at least for me (>13 million rows) and this feature request seems sensible. I thought there would be an option to limit to the first 1000 rows or something but the only thing that is shown is the cache table which I think is for a different purpose. 

When I do a calculation, I want to make sure that this has occurred, but for some reason unless you save the file working with the attribute table becomes impossible as each scroll down is querying something in the back (probably a temporary delta).

@NyakudyaA doing a processing Random Selection would require having the layer already in committed state I suspect, and Open Selected feature requires interacting with millions of points on the map canvas which will be slow also. 

In any case having an option to open just the first 100 rows (better if it was a setting for the number of rows) would be a good improvement

**Comment by theroggy** [comment]
Related: https://github.com/qgis/QGIS/issues/37510
